### ISSUE TITLE
when I  add code hvd.allgather(fc_feature) in tensorFlow_mnist.py, loss can not  converge

### ISSUE THREAD

**Issue by lamperouge11** [horovod/horovod#251]
def conv_model(feature, target, mode):
    """2-layer convolution model."""
    # Convert the target to a one-hot tensor of shape (batch_size, 10) and
    # with a on-value of 1 for each one-hot vector of length 10.
    target = tf.one_hot(tf.cast(target, tf.int32), 10, 1, 0)

    # Reshape feature to 4d tensor with 2nd and 3rd dimensions being
    # image width and height final dimension being the number of color channels.
    feature = tf.reshape(feature, [-1, 28, 28, 1])
   
    with tf.variable_scope('conv_layer1'):
        h_conv1 = layers.conv2d(
            feature, 32, kernel_size=[5, 5], activation_fn=tf.nn.relu)
        h_pool1 = tf.nn.max_pool(
            h_conv1, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')

    # Second conv layer will compute 64 features for each 5x5 patch.
    with tf.variable_scope('conv_layer2'):
        h_conv2 = layers.conv2d(
            h_pool1, 64, kernel_size=[5, 5], activation_fn=tf.nn.relu)
        h_pool2 = tf.nn.max_pool(
            h_conv2, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
        # reshape tensor into a batch of vectors
        h_pool2_flat = tf.reshape(h_pool2, [-1, 7 * 7 * 64])

    # Densely connected layer with 1024 neurons.

    h_fc1 = layers.dropout(
        layers.fully_connected(
            h_pool2_flat, 1024, activation_fn=tf.nn.relu),
        keep_prob=0.5,
        is_training=mode == tf.contrib.learn.ModeKeys.TRAIN)

    # Compute logits (1 per class) and compute loss.

    h_fc1 = hvd.allgather(h_fc1)  ######################### hvd.allgather(h_fc1)
    target = hvd.allgather(target) ########################  hvd.allgather(target)

    logits = layers.fully_connected(h_fc1, 10, activation_fn=None)
    loss = tf.losses.softmax_cross_entropy(target, logits)

    params = [v for v in tf.trainable_variables()]
    grads = tf.gradients(loss, params, aggregation_method=tf.AggregationMethod.DEFAULT)

    return tf.argmax(logits, 1), loss ,grads

**Comment by lamperouge11** [comment]
@jiezhang ,please help me....
no allgather

![image](https://user-images.githubusercontent.com/17671593/39245997-e4f04c70-48c7-11e8-99f9-189c7cd03327.png)
add allgather


![image](https://user-images.githubusercontent.com/17671593/39246017-f7e185b0-48c7-11e8-8d0a-9466c39ba918.png)

**Comment by alsrgv** [comment]
@lamperouge11, did you mean to use `hvd.allreduce` rather than `hvd.allgather`?

**Comment by lamperouge11** [comment]
@alsrgv ,hi , I am sorry about detailed description of the problem。
 I wanted to train a model of great class(more than 500000 classification) within multiple GPU , but the parameter of  the last full_connect  layer is too huge(about 1-2G), so I used to allgather the label and feature before the last full_connect  layer for avoiding transmission parameters , It can compute parameter instead of transmission  parameter on GPU.
As a result, I found the loss can not converge after allgather feature and label .WHY？？？

**Comment by alsrgv** [comment]
@lamperouge11, I see.  I suspect the issue happens because `hvd.allreduce()` and `hvd.allgather()` are marked as not differentiable operations.  What if instead of splitting the layer, you put first layers on GPU:0 and this large layer on GPU:1?  You can do that with `with tf.device("/gpu:0"):` context managers.  You can see #96 for more discussion on the subject.

**Comment by transuranium88** [comment]
@alsrgv I am trying to do the same thing (split the last fc layer onto multiple devices). I saw there is gradient implementation for allgather operation now, so I think this should be supported, right?

**Comment by alsrgv** [comment]
Yes, now they are differentiable.  Please let us know if you see any issues!
